### ISSUE TITLE
Helping people write better / grading readability

### ISSUE THREAD

**Issue by jancborchardt** [nextcloud/text#3]
We could have a mode which helps people write better by rating the readability of their text. This mostly applies if you work on blog posts or essays. I’m imagining something similar to https://hemingwayapp.com, https://www.grammarly.com/, https://quillbot.com/, https://prowritingaid.com/ – etc.

In write mode:
![hemingway-writemode](https://user-images.githubusercontent.com/925062/56140636-9112aa80-5f9b-11e9-8719-ed7ab0b0c2fb.png)

In edit mode (as in "being the Editor"):
![hemingway](https://user-images.githubusercontent.com/925062/56140637-9112aa80-5f9b-11e9-945b-8076529f82ea.png)

**Comment by conloos** [comment]
Hi,

I had in the duplicated ticket: #497 noted that it would be good to design the extension so that the "http-api" from languagetool is used by default and if you have your own server, you can overwrite the destination address.

Thanks Frank

**Comment by juliushaertl** [comment]
> I had in the duplicated ticket: #497 noted that it would be good to design the extension so that the "http-api" from languagetool is used by default and if you have your own server, you can overwrite the destination address.

I don't think we should use their public API by default, since this would mean we transfer user data to a 3rd party without consent. It would be fine as an advanced setting I'd say.

**Comment by jancborchardt** [comment]
Also for reference what word in Microsoft 365 does (more details in the [blog post](https://techcrunch.com/2020/03/30/office-365-becomes-microsoft-365-and-gets-new-personal-and-family-plans/?guccounter=1)):

Direct link to GIF (too large for GitHub embed): https://techcrunch.com/wp-content/uploads/2020/03/Microsoft-Editor-in-Word.gif

**Comment by brainchild0** [comment]
It seems impossible to create an algorithm that satisfies everyone as  a means to superior style  in some particular natural language. Hemmingway had particular ideas about style that many favor, but he is not objectively better than everyone else who has ever written in English.

It is interesting to consider an extensions interface by which the user can select among various choices of interactive style aids. 

I tend to think that exclusive support for a single algorithm will frustrate at least as many users as it will make happy, and I would speculate that some of the most frustrated users will be some of the best writers!

**Comment by jancborchardt** [comment]
Tool found by @monsieurhannes in duplicate issue https://github.com/nextcloud/text/issues/647
https://github.com/kdzwinel/Proofreader

**Comment by brainchild0** [comment]
It seems that Proofreader  is not particularly current, but is anyway just a wrapper for [write-good](https://github.com/btford/write-good), which is maintained actively, adding format support  for Markdown or HTML. Since an editor   maintains an abstract representation of a document, perhaps the latter project would be  of a greater direct use anyway.

**Comment by mgifford** [comment]
I wanted to pass along this in the thread: https://github.com/wooorm/readability

**Comment by sereneinserenade** [comment]
Hi, maybe I'm too late to party, but I've made a Tiptap Extension that integrated spellchecking to Tiptap using ProseMirror decorations and Tiptap BubbleMenu.

Here's the Repository: https://github.com/sereneinserenade/tiptap-languagetool
Here's a demo: https://tiptap-languagetool.vercel.app/

**Comment by mgifford** [comment]
That's pretty neat @sereneinserenade but I don't see how it helps with plain language and improved readability.

Better writing, yes.  Seems like a nice addition to a basic spell-checker.

**Comment by max-nextcloud** [comment]
I think the next step here would be more research in terms of tools we could build upon.

**Comment by brainchild0** [comment]
It might be interesting to consider a pluggable interface allowing users to choose from an assortment of tools by various providers.

While certain style choices may be considered more standard in certain contexts, I push back against the premise that writing may be objectively graded. 

The closest an automated tool might come to helping "write better" would be detecting violations of a very small set of style rules. Again, no single set of guidelines should be imposed on all users.  Even professional (human) editors    understand that their changes impose their own objectives and ideals upon another's work.

**Comment by juliushaertl** [comment]
cc @marcelklehr as I remember seeing some pretrained models for analysing text in terms of agressive wording. Maybe you know something that could be interesting here from the AI perspective.

**Comment by brainchild0** [comment]
> from the AI perspective

Writing isn't chess.   The suggestions given by any AI system of a current level of sophistication would severely degrade the style of a talented writer.

**Comment by mgifford** [comment]
I'm not sure that most people writing content on the web fit the "talented writer" category. I'm also not sure that a really talented writer would be bothered by the suggestions given by AI. After all, we already have grammar check introducing suggestions. Is this making it better or worse? Not sure. AI could also be trained to write in the style of, journalism, a legal document, or Hemingway. No reason why it wouldn't augment a talented writers ability to express themselves in different styles of writing.

**Comment by marcelklehr** [comment]
The latest model from google I mentioned yesterday is quite smart in dealing with queries like ["is the following sentence readable?"](https://huggingface.co/google/flan-t5-large?text=Q%3A+Is+the+following+sentence+easily+readable%3F+%22Ich+kann+nicht+wissen%2C+ob+ich+jemals+gewusst+habe%2C+was+ich+jetzt+wei%C3%9F%2C+nicht+gewusst+zu+haben.%22) For more detailed things like suggestions and per-word analysis I think we won't find a pre-trained model, as even in commercial tools there's probably a lot of hard coding involved.

**Comment by juliusknorr** [comment]
https://writewithharper.com might be interesting

**Comment by hex-m** [comment]
LWN published a [overview](https://lwn.net/Articles/822969/) over possible tools in 2020 and more recently a more in-depth [article about Vale](https://lwn.net/Articles/964075/).

My use-case is producing simple/understandable text as an accessibility measure.
